apps/site/assets/ts/app/channels.ts

```diff
@@ -0,0 +1,95 @@
+import { Socket, Channel } from "phoenix";
+
+declare global {
+  interface Window {
+    channels: { [id: string]: Channel };
+    socket: Socket;
+  }
+}
+
+type UpdateEventName = "reset" | "add" | "update";
+
+interface UpdateEvent<DataType> {
+  event: UpdateEventName;
+  data: DataType;
+}
+
+interface RemoveEvent {
+  event: "remove";
+  data: string[];
+}
+
+export type SocketEvent<DataType> = UpdateEvent<DataType> | RemoveEvent;
+
+const isVehicleChannel = (channelId: string): boolean =>
+  channelId.includes("vehicles:") && channelId !== "vehicles:remove";
+
+const joinChannel = <T>(channelId: string): void => {
+  if (!window.socket) return;
+
+  if (!window.channels[channelId]) {
+    window.channels[channelId] = window.socket.channel(channelId, {});
+  }
+
+  const channel = window.channels[channelId];
+
+  if (!["joined", "joining"].includes(channel.state)) {
+    channel
+      .join()
+      .receive("error", ({ reason }) =>
+        /* eslint-disable no-console */
+        console.error(`failed to join ${channelId}`, reason)
+      )
+      .receive("ok", () => {
+        console.log(`success joining ${channelId}`);
+        /* eslint-enable no-console */
+        if (isVehicleChannel(channelId)) {
+          const [, route_id, direction_id] = channelId.split(":");
```

Comment:
I've actually not seen this syntax before. Is this the equivalent of underscoring an unused value in Elixir?

Reply:
Yep, exactly! It's apparently a valid way to skip assigning variables when destructuring.